lib/full-site-editing/template-loader.php

```diff
@@ -64,57 +64,29 @@ function get_template_hierarchy( $template_type ) {
  */
 function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
 	global $_wp_current_template_content;
-	$current_template = gutenberg_resolve_template( $type, $templates );
-
-	// Allow falling back to a PHP template if it has a higher priority than the block template.
-	$current_template_slug       = str_replace(
-		array( trailingslashit( get_stylesheet_directory() ), trailingslashit( get_template_directory() ), '.php' ),
-		'',
-		$template
-	);
-	$current_block_template_slug = is_object( $current_template ) ? $current_template->slug : false;
-	foreach ( $templates as $template_item ) {
-
-		$template_item_slug = gutenberg_strip_php_suffix( $template_item );
-
-		// Break the loop if the block-template matches the template slug.
-		if ( $current_block_template_slug === $template_item_slug ) {
-
-			// if the theme is a child theme we want to check if a php template exists.
-			if ( is_child_theme() ) {
-
-				$has_php_template   = file_exists( get_stylesheet_directory() . '/' . $current_template_slug . '.php' );
-				$block_template     = _gutenberg_get_template_file( 'wp_template', $current_block_template_slug );
-				$has_block_template = false;
-
-				if ( null !== $block_template && wp_get_theme()->get_stylesheet() === $block_template['theme'] ) {
-					$has_block_template = true;
-				}
-				// and that a corresponding block template from the theme and not the parent doesn't exist.
-				if ( $has_php_template && ! $has_block_template ) {
-					return $template;
-				}
-			}
-
-			break;
-		}
 
-		// Is this a custom template?
-		// This check should be removed when merged in core.
-		// Instead, wp_templates should be considered valid in locate_template.
-		$is_custom_template = 0 === strpos( $current_block_template_slug, 'wp-custom-template-' );
-
-		// Don't override the template if we find a template matching the slug we look for
-		// and which does not match a block template slug.
-		if (
-			! $is_custom_template &&
-			$current_template_slug !== $current_block_template_slug &&
-			$current_template_slug === $template_item_slug
-		) {
-			return $template;
-		}
```

Comment:
So we're iterating over the `$templates` candidates.

1. The slug for the fallback PHP `$template` is in `$current_template_slug`.
2. The slug for the best block template match for `$templates` candidates is in `$current_block_template_slug`.
3. The slug for the current item in the `$templates` iteration loop (without `.php` suffix) is in `$template_item_slug`. 

We return the PHP `$template` (1.) rather than the block template (2.) if:
- The block template part of the regular template hierarchy (and not a [custom page template](https://developer.wordpress.org/themes/template-files-section/page-template-files/)), and
- the block and PHP templates have different slugs -- meaning they apply to different items in the `$templates` candidates list, and
- the PHP template is for the item in the `$templates` candidates list that we're currently iterating over.

Room for optimization mostly comes from the fact that we only want to consider block templates with _higher_ specificity than the candidate PHP `$template` (instead of the inequality check found here; the higher specificity is implicitly handled here by the loop-breaking logic in L81-L100).